blockchain/peer.go

```diff
@@ -0,0 +1,130 @@
+package blockchain
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	flow "github.com/tendermint/tendermint/libs/flowrate"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p"
+	"github.com/tendermint/tendermint/types"
+)
+
+const defaultPeerTimeout = 15 * time.Second
+
+//--------
+// Peer
+var (
+	// Timeout for a peer to respond to a block request.
+	peerTimeout = defaultPeerTimeout
+
+	// Minimum recv rate to ensure we're receiving blocks from a peer fast
+	// enough. If a peer is not sending data at at least that rate, we
+	// consider them to have timedout and we disconnect.
+	//
+	// Assuming a DSL connection (not a good choice) 128 Kbps (upload) ~ 15 KB/s,
+	// sending data across atlantic ~ 7.5 KB/s.
+	minRecvRate = int64(7680)
+
+	// Monitor parameters
+	peerSampleRate = time.Second
+	peerWindowSize = 40 * peerSampleRate
+)
+
+type bpPeer struct {
+	logger log.Logger
+	id     p2p.ID
+
+	height      int64                  // the peer reported height
+	numPending  int32                  // number of requests still waiting for block responses
+	blocks      map[int64]*types.Block // blocks received or expected to be received from this peer
+	timeout     *time.Timer
+	recvMonitor *flow.Monitor
+
+	errFunc func(err error, peerID p2p.ID) // function to call on error
+}
+
+func newBPPeer(
+	peerID p2p.ID, height int64, errFunc func(err error, peerID p2p.ID)) *bpPeer {
+	peer := &bpPeer{
+		id:         peerID,
+		height:     height,
+		blocks:     make(map[int64]*types.Block, maxRequestsPerPeer),
+		numPending: 0,
+		logger:     log.NewNopLogger(),
+		errFunc:    errFunc,
+	}
+	return peer
+}
+
+func (peer *bpPeer) setLogger(l log.Logger) {
```

Comment:
Why it's not public? Same questions for other bpPeer methods. We can keep bpPeer private, but I see no reasons to not make its methods public - at least those what form public API for modifying its state. like incrPending, decrPending, ...

Reply:
I changed them but not convinced as they are internal to the package.